File: lib/codeclimate/collectors/pagerduty/api_client.rb
require "net/http"
require "uri"

module Codeclimate
  module Collectors
    module Pagerduty
      class ApiClient
        BASE_URL = "https://api.pagerduty.com".freeze

        ResponseError = Class.new(StandardError)
        NotFound = Class.new(ResponseError)
        Unauthorized = Class.new(ResponseError)
        ServerError = Class.new(ResponseError)

        def initialize(api_token)
          @api_token = api_token
        end

        def get(path, params = {})
          uri = construct_uri(path, params)
          req = Net::HTTP::Get.new(uri)
          set_headers(req)
          make_request(req)
        end

        private

        attr_reader :api_token

        def make_request(request)
          resp = http_client.request(request)

          case resp
          when Net::HTTPSuccess
            JSON.parse(resp.body)
          when Net::HTTPNotFound
            raise NotFound, "Resource not found"
          when Net::HTTPUnauthorized
            raise Unauthorized, "Invalid credentials"
          when Net::HTTPServerError
            raise ServerError, "Server error"
          else
            raise ResponseError, "Unsuccessful request: status=#{resp.status}"
          end
        end

        def set_headers(req)
          req["Authorization"] = "Token token=#{api_token}"
          req["Content-Type"] = "application/json"
          req["Accept"] = "application/vnd.pagerduty+json;version=2"
        end

        def http_client
          @http_client ||= begin
            base_uri = URI.parse(BASE_URL)

            Net::HTTP.new(base_uri.hostname, base_uri.port).tap do |c|
              c.use_ssl = true
            end
          end
        end

        def construct_uri(path, params = {})
          URI.parse(BASE_URL).tap do |uri|
            puts "construct_uri: initial = #{uri.inspect}"
            if path.start_with?("/")
              uri.path = path
            else
              uri.path = "/#{path}"
            end
            puts "construct_uri: after path set = #{uri.inspect}"

            if params.any?
              uri.query = URI.encode_www_form(params)
            end
          end
        end
      end
    end
  end
end

File: src/ApiClient.ts
import { URL } from "url";
import * as https from "https";

export class ResponseError extends Error {
 constructor(message: string) {
  super(message);
  this.name = ResponseError.name;
  Object.setPrototypeOf(this, ResponseError.prototype);
 }
}
export class NotFoundError extends ResponseError {
 constructor(message: string) {
  super(message);
  this.name = NotFoundError.name;
  Object.setPrototypeOf(this, NotFoundError.prototype);
 }
}
export class UnauthorizedError extends ResponseError {
 constructor(message: string) {
  super(message);
  this.name = UnauthorizedError.name;
  Object.setPrototypeOf(this, UnauthorizedError.prototype);
 }
}
export class ServerError extends ResponseError {
 constructor(message: string) {
  super(message);
  this.name = ServerError.name;
  Object.setPrototypeOf(this, ServerError.prototype);
 }
}

const BASE_URL = "https://api.pagerduty.com";
const VELOCITY_HOST_NAME = "velocity.codeclimate.com";

export class ApiClient {
 constructor(public apiToken: string) {}

 post(path: string, record: any): any {
  const options = {
   hostname: VELOCITY_HOST_NAME,
   port: 443,
   path: path,
   method: "POST",
   headers: {
    Authorization: `token ${this.apiToken}`,
    "Content-Type": "application/vnd.api+json",
   },
  };

  const buildRecord = JSON.stringify({
   data: {
    type: "events",
    attributes: {
     sourcePayload: record,
    },
   },
  });

  const req = https.request(options, (res) => {
   res.on("data", (d) => {
    process.stdout.write(d);
   });
   res.on("error", (err) => {
    console.log("Error: ", err.message);
   });
  });

  req.write(buildRecord);
  req.end();
 }

 // returns a promise that will resolve to a parsed JSON object
 get(path: string, params?: object): Promise<object> {
  return new Promise((resolve, reject) => {
   https
    .get(
     this.resolveUrl(path, params),
     {
      method: "GET",
      headers: {
       Accept: "application/vnd.pagerduty+json;version=2",
       Authorization: `Token token=${this.apiToken}`,
       "Content-Type": "application/json",
      },
     },
     (resp) => {
      if (resp.statusCode && resp.statusCode >= 500) {
       reject(new ServerError("Server error"));
      } else if (resp.statusCode === 401) {
       reject(new UnauthorizedError("Unauthorized"));
      } else if (resp.statusCode === 404) {
       reject(new NotFoundError("Not found"));
      } else if (resp.statusCode !== 200) {
       reject(new ResponseError(`Did not expect ${resp.statusCode} response`));
      }

      let bodyStr = "";

      resp.on("data", (chunk) => (bodyStr += chunk));
      resp.on("end", () => {
       if (resp.statusCode === 200) {
        resolve(JSON.parse(bodyStr));
       }
      });
     }
    )
    .on("error", (err) => {
     reject(new ResponseError(err.toString()));
    });
  });
 }

 private resolveUrl(path: string, params?: object): URL {
  let url = new URL(path, BASE_URL);

  if (params) {
   Object.keys(params).forEach((key) => {
    url.searchParams.set(key, params[key]);
   });
  }

  return url;
 }
}
